docs/design/registries.md

```diff
@@ -0,0 +1,477 @@
+
+### `Cargo.toml` syntax
+
+This design proposes the following tables in `Cargo.toml`:
+
+* `[package.metadata.component]`
+
+  This table contains information about the component being authored. The 
+  supported fields are:
+
+  * `package` - the name of the component package to use when publishing to a 
+    component registry; required to publish.
+
+  * `targets` - a target world for the component being authored; this causes 
+    bindings to be generated for the specified world's imports and exports.
+    
+  * `world` - a path to a wit document specifying a world definition for the 
+    component being authored; it must contain a `world` with the same name as 
+    the current crate name.
+
+  The `targets` and `world` fields are mutually exclusive.
+
+* `[package.metadata.component.dependencies]`
+
+  This table contains information about component dependencies.
+  
+  The dependencies may come from a component registry or from local wit 
+  documents and components.
+
+  Specifying dependencies from a registry is done in the general form of:
+
+  ```toml
+  name = { package = "<package-id>", version = "<version>", registry = "<registry>" }
+  ```
+
+  The `registry` field is optional and defaults to `default` (see the `[package.metadata.component.registries]`
+  table below).
+
+  As packages in a component registry are namespaced (unlike Rust crate 
+  registries), the short-form differs from what is supported by `cargo`:
+
+  ```toml
+  name = "<package-id>:<version>"
+  ```
+
+  Which is equivalent to:
+
+  ```toml
+  name = { package = "<package-id>", version = "<version>" }
+  ```
+
+  Local wit documents and components are specified using the `path` field:
+
+  ```toml
+  name = { path = "<path>" }
+  ```
+
+  In the future, it may be possible to specify a path to a directory containing 
+  a `Cargo.toml` that itself defines a component and treat it as a component 
+  package dependency.
+
+  The names of dependencies correspond to "packages" in a `wit` document and 
+  may also influence the names of imports in the component being authored ([see dependency behavior](#dependency-behavior)).
+
+* `[package.metadata.component.registries]`
+
+  This table is entirely optional and is a map of friendly names to registry 
+  URL (the format of the URL is dependent upon the underlying 
+  registry client implementation):
+
+  ```toml
+  name = "<registry-url>"
+  ```
+
+  Dependencies may specify a specific registry to use by specifying the `registry` field:
+
+  ```toml
+  foo = { package = "ns/foo", version = "0.1.0", registry = "my-registry" }
+  ```
+
+  The `default` name is used when no `registry` field is specified in a
+  dependency. Therefore, specifying a registry with name `default` will 
+  override the built-in default in `cargo-component` (expected to be a 
+  future Bytecode Alliance component registry instance).
+
+  A registry URL may use the `file://` scheme to point at local directory that 
```

Comment:
We probably want to use some other format for a file-system based registry as the `file` URL scheme can't express immediate relative paths (e.g. `file://foo` isn't technically allowed).

Perhaps instead express a registry entry in the form `name = { path = "<path>" }`.

Thus the form `name = "<registry-url>"` is just a short-hand for `name = { url = "registry-url" }`.